pretyping/unification.ml

```diff
@@ -1960,64 +1961,17 @@ let w_unify_to_subterm env evd ?(flags=default_unify_flags ()) (op,cl) =
        else user_err Pp.(str "Bound 1")
      with ex when precatchable_exception ex ->
        (match AConstr.kind cl with
-          | App (f,args) ->
-              let n = Array.length args in
-              assert (n>0);
-              let c1 = AConstr.mkApp (f,Array.sub args 0 (n-1)) in
-              let c2 = args.(n-1) in
-              (try
-                 matchrec c1
-               with ex when precatchable_exception ex ->
-                 matchrec c2)
-          | Case(_,_,_,_,_,c,lf) -> (* does not search in the predicate *)
-               (try
-                 matchrec c
-               with ex when precatchable_exception ex ->
-                 iter_fail matchrec (Array.map snd lf))
-          | LetIn(_,c1,_,c2) ->
-               (try
-                 matchrec c1
-               with ex when precatchable_exception ex ->
-                 matchrec c2)
-
-          | Proj (p,c) -> matchrec c
-
-          | Fix(_,(_,types,terms)) ->
-               (try
-                 iter_fail matchrec types
-               with ex when precatchable_exception ex ->
-                 iter_fail matchrec terms)
-
-          | CoFix(_,(_,types,terms)) ->
-               (try
-                 iter_fail matchrec types
-               with ex when precatchable_exception ex ->
-                 iter_fail matchrec terms)
-
-          | Prod (_,t,c) ->
-              (try
-                 matchrec t
-               with ex when precatchable_exception ex ->
-                 matchrec c)
-
-          | Lambda (_,t,c) ->
-              (try
-                 matchrec t
-               with ex when precatchable_exception ex ->
-                 matchrec c)
-
-          | Array(_u,t,def,ty) ->
-            (try
-              matchrec def
-            with ex when precatchable_exception ex ->
-             try
-              matchrec ty
-            with ex when precatchable_exception ex ->
-              iter_fail matchrec t)
-
-          | Cast (_, _, _) (* Is this expected? *)
-          | Rel _ | Var _ | Meta _ | Evar _ | Sort _ | Const _ | Ind _
-            | Construct _ | Int _ | Float _ -> user_err Pp.(str "Match_subterm")))
+        | ACast _ -> assert false (* just got stripped *)
+        | AApp (f,args) ->
+          let n = Array.length args in
+          assert (n>0);
+          let c1 = AConstr.mkApp (f,Array.sub args 0 (n-1)) in
```

Comment:
With this simpler representation, we could even try to be a bit more clever and not reconstruct the subarrays here.